changes to dao.js and memory.js to add a new method on Upsert based on non-primary key

mountain1234585 opened the pull request:
Hi Raymond,
Kindly review the basic changes done to loopback-datasource-juggler lib/dao.js and lib/connectors/memory.js to include a new method called "upsertWithWhere" based on non-primary key.
1)This method leverages the existing methods, find, create, findById and updateAttributes.
2)I have not added notifyObserversOf like in other methods , since this method calls them internally.
3)Once you review the code changes, we will add the unit tests as well.
4) I have tested the operation manually by creating an application, memory db datasource and persisted model in case of create and update scenarios.
Thanks,
Sonali Samantaray

slnode commented:
Can one of the admins verify this patch? To accept patch and trigger a build add comment ".*ok\W+to\W+test.*"

slnode commented:
Can one of the admins verify this patch?

slnode commented:
Can one of the admins verify this patch?

slnode commented:
Can one of the admins verify this patch?

raymondfeng commented on lib/connectors/memory.js:
```diff
@@ -792,6 +792,105 @@ Memory.prototype.replaceOrCreate = function(model, data, options, callback) {
   });
 };
 
+Memory.prototype.upsertWithWhere = function(model,data,options,callback){
```
The method signature is not correct. The `where` argument is missing.

raymondfeng commented on lib/connectors/memory.js:
```diff
@@ -792,6 +792,105 @@ Memory.prototype.replaceOrCreate = function(model, data, options, callback) {
+      response = true;
+
+    if (!data.hasOwnProperty(element)) {
+      data[element] = where[element];
```
For `upsertWithWhere`, we should not merge things from `where` into `data`. The `where` argument should be the matching criteria while the `data` argument should be the properties to be inserted or updated.

raymondfeng commented on lib/connectors/memory.js:
```diff
@@ -792,6 +792,105 @@ Memory.prototype.replaceOrCreate = function(model, data, options, callback) {
+    var filter = {
+      "where": where
+    };
+
```
There are too many blank lines.

Amir-61 replied:
Please run our linter; I see lots of lint issues

raymondfeng commented on lib/connectors/memory.js:
```diff
@@ -792,6 +792,105 @@ Memory.prototype.replaceOrCreate = function(model, data, options, callback) {
+
+        if (err) {
+
+          callback(err);
```
Please wrap `callback` into `process.nextTick()` so that callback only happens after the function is finished.

raymondfeng commented on lib/connectors/memory.js:
```diff
@@ -792,6 +792,105 @@ Memory.prototype.replaceOrCreate = function(model, data, options, callback) {
+
+          if (models.length > 1) {
+
+            callback("Error message:" + "Multiple records exists.Upsert can't be performed.");
```
Why don't we update multiple matching records?

raymondfeng commented on lib/dao.js:
```diff
@@ -68,8 +67,8 @@ function convertSubsetOfPropertiesByType(inst, data) {
     typedData[key] = inst[key];
     if (typeof typedData[key] === 'object' &&
-            typedData[key] !== null &&
-            typeof typedData[key].toObject === 'function') {
```
There are a lot of style changes mixed with functional changes. Please either keep the existing style or commit the style changes in a separate one.

raymondfeng commented on lib/dao.js:
```diff
@@ -2783,192 +2836,308 @@ DataAccessObject.replaceById = function(id, data, options, cb) {
-    }
-  }
+  var element = Object.keys(where)[0];
+  console.log(element);
```
Please remove console.log or turn it into a debug statement.

raymondfeng commented on lib/dao.js:
```diff
@@ -2783,192 +2836,308 @@ DataAccessObject.replaceById = function(id, data, options, cb) {
+
+else
+{
+cb("data field can't be empty");
```
Bad style

slnode commented:
Can one of the admins verify this patch?

slnode commented:
Can one of the admins verify this patch?

slnode commented:
Can one of the admins verify this patch?

Amir-61 commented on lib/connectors/memory.js:
```diff
@@ -792,6 +792,105 @@ Memory.prototype.replaceOrCreate = function(model, data, options, callback) {
+            self.create(data, function (err, results) {
+
+              if (err)
+                callback(err);
```
We always use:

```
if (err) callback(err);
```

Amir-61 commented on lib/connectors/memory.js:
```diff
@@ -792,6 +792,105 @@ Memory.prototype.replaceOrCreate = function(model, data, options, callback) {
+            self.findById(idVal, function (err, modelInst) {
+              if (err) {
+
+                callback(err);
```
ditto

Amir-61 commented on lib/connectors/memory.js:
```diff
@@ -792,6 +792,105 @@ Memory.prototype.replaceOrCreate = function(model, data, options, callback) {
+
+                  if (err) {
+
+                    callback(err);
```
ditto; please fix in other places as well

Amir-61 commented on lib/connectors/memory.js:
```diff
@@ -792,6 +792,105 @@ Memory.prototype.replaceOrCreate = function(model, data, options, callback) {
+    callback(ex);
+  }
+
+
```
I see too many blank lines; please fix them

Amir-61 commented on lib/dao.js:
```diff
@@ -83,7 +82,8 @@ function convertSubsetOfPropertiesByType(inst, data) {
   var keys = Object.keys(data);
-  var result = {}, key;
+  var result = {},
+    key;
```
Please avoid changing un-related code in this PR

Amir-61 commented on lib/dao.js:
```diff
@@ -106,11 +106,14 @@ function setIdValue(m, data, value) {
-  var query = { where: {}};
+  var query = {
+    where: {}
+  };
```
Please dont change un-related code in this PR

Amir-61 commented:
@mountain1234585 
Did you sign [CLA](https://cla.strongloop.com/agreements/strongloop/loopback-datasource-juggler)? Please add tests and run npm test locally and fix issues as well as lint issues. You can type `@slnode test please` to run tests on CI to make sure you are passing tests.

mountain1234585 commented:
Hello Raymond,

Thanks for your detailed review in a short span of time.

Following on the comments you had on the logic :

1) why we don't update multiple instances rather return error if multiple instances are found.
	This was discussed as part of the requirement we had and we plan to restrict this only for single instance update. I had a look at how upsert with foreign key is performed in some of the popularly used endpoints and they follow the similar approach.  

2) why the key value pair mentioned in the "where" is merged into data
        In case of create, we want the field to be updated when a new record is created. And similar is seen in upsert operation with Salesforce as well.

3) method signature is wrong in memory.js
	I have now removed the method in memory.js since the entire logic is handled in dao.js
4) too many empty lines in dao.js
	I have removed them now.


Regards,
Sonali Samantaray

mountain1234585 commented:
@Amir-61
I have fixed the elint errors now.

mountain1234585 commented:
I think I need to take the latest code, since I can see some commits that has happened in between.

mountain1234585 commented:
There are too many conflicts in this PR, while trying to rebase and sync. I am doing these changes and have submitted a PR as part of PR https://github.com/strongloop/loopback-datasource-juggler/pull/1001
